Site Editor - add query args for current context.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
Initial stab at #22238.  This PR allows us to load the site editor to a specific context via query args `postType` and `postId`

Based on @vindl's comment [here](https://github.com/WordPress/gutenberg/pull/27124#discussion_r531797052), this PR has been changed from the original implementation.

The idea is that `postType` can be one of: "page", "post", "wp_template", "wp_template_part". Then the ID would be whichever post we mean to display. This covers several scenarios:

1. If we wish to link directly to a "page", we use `postType=page&postId=2`. This is enough information for the site editor to resolve the correct template to display. Same for "post"
2. If we wish to link directly to a template with no content, we can use `postType=wp_template&postId=2`. This will load the specified template without any page context set. Same for "wp_template_part".

Limitations:
1. Does not support other context types yet, such as "category" or anything related to queries.
2. If we need to resolve templates or template parts by slugs if there is no post ID, this will not work.

I think this is simple enough for a round one without having to cover the edge cases by serializing the JSON state into the URL. What do you think?

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Load the site editor as you would normally.
* select another entity in the nav panel, verify query args are added to the url.
* reload the browser, verify the editor loads to entity corresponding to the query args.

Verify this works for various contexts including: categories, pages, posts, templates, and template parts.
Verify creating a new template updates the url as well.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
Use query args for routing

Addison-Stavlo commented:
One small question I have regarding this is about how we would like the browser's back button to function.  Say a user enters the site editor through wp-admin, navigates between a variety of entities using the nav panel, and finally selects "back" on the browser history, should this:

* Cycle back through all the entities inside the editor that they have navigated across.

or 

* Go straight back to wp-admin.

Technically either way seems simple enough, but I wonder what makes the most sense for an editor like this. 🤔

ntsekouras commented on packages/edit-site/src/components/post-router/update-query-params.js:
```diff
@@ -0,0 +1,23 @@
+	const queryParams = new URLSearchParams( window.location.search );
+	if ( contextType === 'content' ) {
+		queryParams.delete( 'id' );
+		queryParams.set( 'contextType', contextType );
```
This is a common statement and could be outside the checks.

noahtallen replied:
Fixed

david-szabo97 commented on packages/edit-site/src/components/post-router/index.js:
```diff
@@ -0,0 +1,61 @@
+		}
+	}, [] );
+
+	// Upadte URL when context changes.
```
```suggestion
	// Update URL when context changes.
```

ntsekouras reviewed (commented):
Thanks for working on this! I believe this will be useful for detecting the context for Query block as well.

This seems to be in a good direction but I need to test it more. I noticed a couple of things for start: 
1. In FSE navigation(sidebar) the description displays html tags. In `templates-> index` description.
2. Having `content` as object in url makes it unreadable. I don't have something to propose yet but will look in more depth in general.

david-szabo97 reviewed (approved):
Love this! Thank you for working on it! ✅ 🙇

gziolo commented on packages/edit-site/src/components/post-router/index.js:
```diff
@@ -0,0 +1,61 @@
+		const id = query.get( 'id' );
+		const content = query.get( 'content' );
+
+		if ( 'content' === contextType && content ) {
```
Linter would catch it if we would care.

gziolo commented on packages/edit-site/src/components/post-router/update-query-params.js:
```diff
@@ -0,0 +1,23 @@
+ * for various content contexts.
+ */
+export default function updateQueryParams( contextType, identifier ) {
+	const queryParams = new URLSearchParams( window.location.search );
```
[`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) – it looks like it isn't supported in IE at all, so it should be polyfilled.

There is also [`@wordpress/url`](https://github.com/WordPress/gutenberg/tree/master/packages/url) package that is used in other places and it has some WordPress specific methods. I have no idea if it's applicable here, but I though I would share.

david-szabo97 replied:
Good point. `getQueryArg` and `addQueryArgs` should be enough for this use-case.

noahtallen replied:
Good idea, switched to use @wordpress/url for all the use cases in this PR

noahtallen committed:
Use @wordpress/url instead

noahtallen commented:
In 8468ac7, I switched to use @wordpress/url instead of the build-in methods. I think with these changes all PR feedback should be addressed! It seems to be working pretty well.

noahtallen commented:
Might need to update some e2e tests, but rerunning the failing ones first to make sure.

noahtallen commented:
I verified that the e2e failure started happening elsewhere before this PR

vindl commented on packages/edit-site/src/components/post-router/index.js:
```diff
@@ -0,0 +1,63 @@
+	// Update URL when context changes.
+	useEffect( () => {
+		if ( page ) {
+			updateQueryParams( 'content', JSON.stringify( page ) );
```
I'm not sure if we should pass all of the page object details into the URL. It seems to me that most of the data there is redundant and can be inferred by post id. For example:

```js
{"type":"page","slug":"sample-page","path":"/?page_id=2","context":{"postType":"page","postId":2}}
```

All of the features outside of editor that want to integrate this will have to conform with this format. They will become and every change to this might break all of the integrations. I think we should try to only pass necessary data as separate URL params, instead of URL encoding an exiting object.

Addison-Stavlo replied:
Good Points.  My main reason for using the entire page object was so we could easily support every editor state, including taxonomies/categories/etc.  We can still develop support for these without setting the entire page object as the value, but we I think we will need to add more individual query args in that case and rebuild the page object from them.

noahtallen committed:
Use new URL format

noahtallen committed:
Remove uneeded changes

ntsekouras commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -196,6 +197,7 @@ function Editor() {
 
 	return (
 		<>
+			<PostRouter />
```
I'm not sure if this naming is very helpful. as in my mind this is something like FSERouter or SiteEditorRouter... It can be addressed later though but wanted to get some thoughts from others..

Addison-Stavlo replied:
Yeah, I don't have any strong reasons for choosing `PostRouter` other than lack of creativity.  `EditorRouter` or something along the flavors you mentioned may be more fitting.

Addison-Stavlo replied:
Or would `QueryController`, or `URLQueryController` make more sense?  Any thoughts?

ntsekouras commented on packages/edit-site/src/store/actions.js:
```diff
@@ -125,6 +126,9 @@ export function setHomeTemplateId( homeTemplateId ) {
  * @return {number} The resolved template ID for the page route.
  */
 export function* setPage( page ) {
+	if ( ! page.path && page.context?.postId ) {
```
Why do we need this?

noahtallen replied:
This `setPage` action tries to find the template which would be used to render the given page via the page path property like this: `example.com/path/to/page?_wp_find_template`. This basically relies on having a path to go to, and then once the template is resolved for that path, the request returns some JSON for the template instead of an HTML page like it normally would.

When we set pages via query args (`?postType=page&postId=2`), we don't have a path set. This means that the `setPage` action wouldn't receive a page path, and therefore would not be able to resolve the template which would be used to render that page.

So to get around that, if the path does not exist, but a post ID does exist, then we use the `?p=2` shorthand as the path so that it can resolve the template.

ntsekouras reviewed (approved):
I tested this and works well for now. I think we can merge and continue on follow ups, especially for handling `taxonomies`.  The change to not having in URL all the `context` seems right to me and makes it readable as well. Will need a rebase for green CI though (thanks to @noahtallen ) :) 

Thanks for working on this 👍 !

Addison-Stavlo commented on packages/edit-site/src/store/reducer.js:
```diff
@@ -139,12 +139,7 @@ export function templateType( state = 'wp_template', action ) {
-		case 'SET_TEMPLATE':
-			if ( action.templateId === state.templateId ) {
-				return state;
-			}
```
I think these changes regarding `SET_TEMPLATE` and the page state structure are still necessary, however they fixed a bug that is present regardless of our query arg changes here.  Consider the flow where:

* A user sets a page to edit.  The editor loads the page and corresponding template.
* The user is done editing the page and wants to go edit a different template.  The user navigates to this template and selects it in the nav sidebar.
* Without this check, the page context will still be set for that unrelated template.

Specifically you can see this bug now if you load the editor (loads on front-page template), then select the sample page, then select the front-page template again you will see the sample page's content in the front-page's post content block.

This shouldn't block the PR though, we can reintroduce this bug fix separately or in the follow up for the taxonomy.

noahtallen replied:
I'm ok following up on it. I wasn't sure exactly how to handle this situation. I could see it being good to keep the page context around -- then you can preview it in different templates. But I can see how that could also be a bug 🤔

Addison-Stavlo replied:
Yeah it is a bit confusing in that regard.  I would argue that switching back to a specific template should clear the page context b/c at that point our context is template editing.  

> I could see it being good to keep the page context around -- then you can preview it in different templates. 

I think we need a separate new mechanism for changing / previewing a different template for the currently selected content.  This current flow is definitely a hacky way to preview that stems from outdated state and doesn't allow the user to actually set anything.

noahtallen replied:
Yeah, that's a good point. I really would like to avoid putting the template ID in the page state. I guess it doesn't _really_ matter, but it seems weird to have to store it twice

Addison-Stavlo replied:
>  it seems weird to have to store it twice

True.  I guess one corresponds to which template is loaded in the editor, while the other corresponds to which is resolved for the page/content.  Maybe the latter should added to the page object itself as `resolvedTempalate` or something similar.  Or maybe there is a better way to handle this.

noahtallen replied:
I like the resolved template idea!

Addison-Stavlo replied:
We will have to look into that.  Anyways, il mark this convo as 'resolved' since I don't want to hold up this PR on an unrelated tangent conversation. 🤣

Addison-Stavlo reviewed (commented):
> Having `content` as object in url makes it unreadable. I don't have something to propose yet but will look in more depth in general.

Agreed, however I was thinking it would be preferable to have an unreadable value for one of these over a simplification that only supports a subset of editor states.  However, I think we can still have the best of both worlds.  

> continue on follow ups, especially for handling `taxonomies`.

We should still be able to support taxonomies without that unreadable value by adding a few more readable query args and using them to rebuild the page object we set on state.  Or are there any other ideas on how to handle them? 🤔   Either way, I also think that will make sense as a follow up.

<hr/>

Thanks for working on this one @noahtallen !  Its looking good 😁

ntsekouras commented:
Is something blocking this PR?

Addison-Stavlo commented:
> Is something blocking this PR?

Nope, sorry!  We have been distracted with some unrelated duties this week. 😢   Il rebase.

Addison-Stavlo committed:
change name of component

Addison-Stavlo commented:
Rebased and updated the name.  Just need tests to pass and a 👍  on the new component name and this should be good to merge.  (also feel free to change it if you think EditorRouter or SiteEditorRouter would still be more fitting).

noahtallen commented on packages/edit-site/src/components/url-query-controller/index.js:
```diff
@@ -0,0 +1,80 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { getQueryArg, addQueryArgs, removeQueryArgs } from '@wordpress/url';
+
+export default function PostRouter() {
```
```suggestion
export default function URLQueryController() {
```

Addison-Stavlo replied:
Whoops!  I even thought to myself that I needed to update that and then totally forgot.  🤣   thanks for fixing that!

noahtallen committed:
Rename PostRouter component function
